Fix Smooth Takeoff

MaEtUgR opened the pull request:
**Describe problem solved by the proposed pull request**
No takeoff in no flight task is smooth anymore.
fixes #12000 when it's finished

**Describe your preferred solution**
All tasks need to be able to take off with the smooth takeoff ramp.
The important changes besides some refactoring are:
- Start from a velocity setpoint pushing into the ground to ramp from idle thrust up.
- Trigger takeoff with a bit higher velocity setpoint threshold to make sure the vehicle has a chance to really get off the ground during the ramp.
- Calculate ramp slope from initialization setpoint to the desired one instead from zero to the desired. This ramps up correctly/quicker when you demand a very small upwards velocity like the AutoLineSmoothVel and ManualPositionSmoothVel tasks do at the beginning.
- Don't stay in the takeoff ramp depending on the land detector, this is unnecessary and can result in problems.

**Additional context**
#11399

RomanBapst commented:
@MaEtUgR Cool, thanks for looking into this. Let me know when you want a review.

MaEtUgR commented:
@RomanBapst Thanks for the offer. I brought the basic functionality back and corrected some details such that it also works with the ...SmoothVel tasks. The important changes are in the third commit and the explanation in the commit message. Feel free to ask back.

I tested in SITL which shows me that it does what it should but not how the real world takeoff user experience is:
https://logs.px4.io/plot_app?log=7bef31f8-0d8e-433e-93d4-134355e5d1a8
https://logs.px4.io/plot_app?log=af0a7e32-ee41-4623-bf70-8345cdecc833

MaEtUgR commented:
What I actually spent most time debugging is a probably in reality very rare corner case where smooth takeoff is skipped:
- you are in manual mode (e.g. after boot)
- you have RC but no mode switch configured
- you have your throttle stick high
- you trigger an auto takeoff from QGC
In this case the smooth takeoff is completely skipped because:
- the attitude controller publishes a high thrust from your stick even while disarmed
- the position controller does not publish any thrust setpoint because no task is running while in manual (because it doesn't want to interfere attitude setpoints) and while disarmed (such that the task gets initialized in the right moment)
- the land detector sees high thrust shortly after arming
- when the task starts running after arming the position controller never saw landed so it skips the smooth takeoff
Possible hotfix for this here: https://github.com/PX4/Firmware/compare/hotfix-manual-high-thrust-on-takeoff

MaEtUgR commented:
@PX4/testflights The goal of this PR is to have a good user experience when taking off from the ground in all possible position and altitude controlled modes.

- Position/Altitude: Having the throttle stick in the middle or below it should stay on the ground in idle. When raising the throttle above half it should not instantly jump but ramp up at the beginning. In position mode it should work with all possible `MPC_POS_MODE` options (especially "smooth" and "smooth velocity").
- Auto takeoff from tablet: should not jump up at the beginning with `MPC_AUTO_MODE` on "default line tracking" or "jerk-limited trajectory"

MaEtUgR commented:
@bresch Thanks for your analysis!
As a side note: There are always two sample points very close together in your log plot, same was true during my log analysis on the weekend and I realized today that when I printf the takeoff ramp there are always two values very close to each other and if something triggers it's always an even number of loop cycles later. Is that a scheduling problem? FYI @julianoes

jorge789 commented:
Tested on PixRacer V4:

Modes Tested:
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.

Procedure:
Arm in position mode, check take off behavior (Vehicle should stay on the ground as long as the throttle stick is at the middle or below, and start to ascend smoothly as long as throttle stick as above the middle.), the land and disarm.
Same procedure was repeated with Altitude and Stabilized mode.

Notes:
No issues were noted on position and altitude, but while testing stabilized mode, vehicle started to ascend slowly before reaching middle position, good flight in general.

Flight log on Position:
https://review.px4.io/plot_app?log=de7cbb9d-03e6-4294-8f63-d6395f0551a3

Flight log Altitude:
https://review.px4.io/plot_app?log=834e6b0c-eb12-4dba-9f3a-715bc1d05c9b

Flight log Stabilized:
https://review.px4.io/plot_app?log=47a7b939-f9e5-4622-af18-3e90700fe7d6

Tested on PixRacer V4

Modes Tested

Takeoff: Good
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

Pre-flight Procedure:

Flash firmware through QGC.OK
Arm vehicle. OK
Calibrations. OK
Parameter check. OK

Flight Test: 
Arm and Take off in position mode. OK
Throttle Test. OK
Pitch, Roll, Yaw. OK
Mode Switch. OK
RTL. OK

Notes:

Took off in Position mode, after flying for approximately one minute, switched to Altitude Mode, then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoints activate RTL and see landing behavior.No issues noted good flight in general. The vehicle behaved as expected. 

Flight log FULL Flight test:
https://review.px4.io/plot_app?log=7adce484-e6ce-4537-804f-0b386092032b

Junkim3DR commented:
### Tested on Pixhawk Pro v4Pro

**Modes Tested:**
- Position Mode: Good. 
- Altitude Mode: Good. 
- Stabilized Mode: Good. 

**Procedure:**
Arm in position mode, check take off behavior (Vehicle should stay on the ground as long as the throttle stick is at the middle or below, and start to ascend smoothly as long as throttle stick as above the middle.), the land and disarm.
Same procedure was repeated with Altitude and Stabilized mode.

**Notes:**
No issues were noted on position and altitude, but while testing stabilized mode, vehicle started to ascend slowly before reaching middle position, good flight in general.

**Logs:**
- Position: https://review.px4.io/plot_app?log=9a058320-300e-40a2-92d8-b1bd1745a095

- Altitude: https://review.px4.io/plot_app?log=042baba5-87df-454a-9e82-6fb2a144d640

- Stabilized: https://review.px4.io/plot_app?log=a27442ae-7df4-4efc-aeb3-11faa2a5dd80

- Full flight test: https://review.px4.io/plot_app?log=ccff6681-180d-4faf-8c0b-ea02eb288579


### Tested on Pixhawk 4Mini v5

**Modes Tested:**
- Position Mode: Good. 
- Altitude Mode: Good. 
- Stabilized Mode: Good. 

**Procedure:**
Arm in position mode, check take off behavior (Vehicle should stay on the ground as long as the throttle stick is at the middle or below, and start to ascend smoothly as long as throttle stick as above the middle.), the land and disarm.
Same procedure was repeated with Altitude and Stabilized mode.

**Notes:**
No issues were noted on position and altitude, but while testing stabilized mode, vehicle started to ascend slowly before reaching middle position, good flight in general.

**Logs:**
- Position: https://review.px4.io/plot_app?log=cd5cfedf-3a31-4309-8040-9fe21083c095

- Altitude: https://review.px4.io/plot_app?log=ec798ed5-f755-4d3a-82cc-ca67bbed12ae

- Stabilized: https://review.px4.io/plot_app?log=ba3c82a6-4e97-40d7-91db-24edf80a228a

- Full flight test: https://review.px4.io/plot_app?log=560fd8dc-131e-4be4-ab3b-4670f324543a

dannyfpv commented:
Tested on Pixhawk4 v5
Modes Tested:

Position Mode: no issue
Altitude Mode: no issue
Stabilized Mode: no issue
Procedure:
Arm in position mode, check take off behavior (Vehicle should stay on the ground as long as the throttle stick is at the middle or below, and start to ascend smoothly as long as throttle stick as above the middle.), the land and disarm.
Same procedure was repeated with Altitude and Stabilized mode.

Notes:
No issues were noted on position and altitude, but while testing stabilized mode, vehicle started to ascend slowly before reaching middle position, good flight in general.

Logs:

Stabilized:  https://review.px4.io/plot_app?log=c624835e-4141-4564-a706-72b75749af5c

Altitude: https://review.px4.io/plot_app?log=18b11371-0ef8-4f68-bd23-e44a9caf7f4c

Position:  https://review.px4.io/plot_app?log=01c85f31-82a4-4e88-a710-e4d399b6ea0f

Full flight test: https://review.px4.io/plot_app?log=7386dd11-f307-4ad2-b614-93e477695c91

Stifael commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -102,8 +102,11 @@ class MulticopterPositionControl : public ModuleBase<MulticopterPositionControl>
+	// smooth takeoff vertical velocity ramp state
+	bool _in_takeoff_ramp = false; /**< true while takeoff ramp is applied */
+	bool _position_triggered_takeoff = false; /**< true if takeoff was triggered by a position setpoint */
+	float _takeoff_ramp_velocity = -1.f; /**< current value of the smooth takeoff ramp */
```
Just a note and not more: to me `velocity` agrees with the coordinate convention (vs speed, which is always positive). Hence, with that convention in mind, `_takeoff_ramp_velocity` of `-1` means that it would start with 1m/s upward.

MaEtUgR replied:
Good point, I should have switched to the proper coordinate convention. Noted to do in the next iteration since this is a release blocker. 👍

MaEtUgR replied:
When #12014 is finished the coordinate frame will be correct.

mhkabir commented:
Can someone test takeoffs in Offboard before this is merged?

dagar commented:
> Can someone test takeoffs in Offboard before this is merged?

@okalachev is that someone you could test quickly?

dagar commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -102,8 +102,11 @@ class MulticopterPositionControl : public ModuleBase<MulticopterPositionControl>
+	bool _in_takeoff_ramp = false; /**< true while takeoff ramp is applied */
+	bool _position_triggered_takeoff = false; /**< true if takeoff was triggered by a position setpoint */
+	float _takeoff_ramp_velocity = -1.f; /**< current value of the smooth takeoff ramp */
+	float _takeoff_reference_z; /**< z-position when takeoff ramp was initiated */
```
```suggestion
	float _takeoff_reference_z{0.0f}; /**< z-position when takeoff ramp was initiated */
```

MaEtUgR commented:
So according to my tests this solution is acceptable, the smooth takeoff works but it's not perfect. The remaining issues for the ramp itself (which also existed before) are:
- Depending on the vehicle tuning the ramp could start a bit above zero thrust.
- If the task that takes off sets a position and a velocity at the same time it can happen that the position setpoint results in a feed forward velocity setpoint addition and there's a small twitch when the takeoff ends.
I would say the minimum viable takeoff is there and I'll work on fixing the remaining issues on https://github.com/PX4/Firmware/pull/12014

MaEtUgR commented:
@dagar I fixed the zero initialization of your review comment.

I also added the hotfix for the problem I described in the comment above: https://github.com/PX4/Firmware/pull/12001#issuecomment-491539137 because I realized it's not a rare corner case but also happened to me in the real world testing.

- Springloaded throttle
- RC already connected on boot
- Vehicle boots in manual
- Switches to position according to my mode switch position
- When I arm it jumps up without takeoff

I had the fix on my testing branch from that moment on and that showed me it's necessary for the release. I discussed with @bkueng, @RomanBapst (who found this exact problem before) and @bresch how to solve the issue and that the hotfix is viable because a proper solution requires more involved position controller changes.

MaEtUgR commented:
I went out testing again and am honestly not very happy with the takeoff user experience. It works but with the default velocity controller tuning the ramp starts too high in my tests and I also see this in the logs of the test team.

Here are some of my logs:
https://logs.px4.io/plot_app?log=deac3730-8462-45dd-b95c-117889a5a1aa
https://logs.px4.io/plot_app?log=6bca7d26-8e30-4a6c-a92d-2c486fdfdb10
https://logs.px4.io/plot_app?log=9feb5493-535c-46ba-98c1-c3b66c0ec40e
https://logs.px4.io/plot_app?log=23786527-9e9f-4bd8-9c5f-454117cb28f0
https://logs.px4.io/plot_app?log=df6fbeef-4015-4245-86d7-d0db940b07d9
https://logs.px4.io/plot_app?log=5e5d19fe-661f-469a-aa4a-d69da50e843a
https://logs.px4.io/plot_app?log=16b7ade7-1742-4e9f-be6b-7613fa29e60e

LorenzMeier commented:
@MaEtUgR What is the next action item here? @bresch @RomanBapst Any ideas you want to contribute?

MaEtUgR commented:
Should have updated: #12014 is the follow up, it contains these changes + all the additional fixes. It's a different pr because initially I thought these minimal quick changes are enough to just fix what we had before but that turned out to be not the case.
